core/executor/src/wasm_runtimes_cache.rs

```diff
@@ -0,0 +1,328 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implements a cache for pre-created Wasm runtime module instances.
+
+use crate::error::Error;
+use crate::wasm_executor::WasmExecutor;
+use log::{trace, warn};
+use parity_codec::Decode;
+use parity_wasm::elements::{deserialize_buffer, DataSegment, Instruction, Module as RawModule};
+use primitives::storage::well_known_keys;
+use primitives::Blake2Hasher;
+use runtime_version::RuntimeVersion;
+use state_machine::Externalities;
+use std::collections::hash_map::{Entry, HashMap};
+use std::mem;
+use std::rc::Rc;
+use wasmi::{Module as WasmModule, ModuleRef as WasmModuleInstanceRef, RuntimeValue};
+
+#[derive(Debug)]
+enum CacheError {
+	CodeNotFound,
+	ApplySnapshotFailed,
+	InvalidModule,
+	CantDeserializeWasm,
+	Instantiation,
+}
+
+/// A runtime along with its version and initial state snapshot.
+#[derive(Clone)]
+pub struct CachedRuntime {
+	/// A wasm module instance.
+	instance: WasmModuleInstanceRef,
+	/// Runtime version according to `Core_version`.
+	///
+	/// Can be `None` if the runtime doesn't expose this function.
+	version: Option<RuntimeVersion>,
+	/// The snapshot of the instance's state taken just after the instantiation.
+	state_snapshot: StateSnapshot,
+}
+
+impl CachedRuntime {
+	/// Perform an operation with the clean version of the runtime wasm instance.
+	pub fn with<R, F>(&self, f: F) -> R
+	where
+		F: FnOnce(&WasmModuleInstanceRef) -> R,
+	{
+		self.state_snapshot.apply(&self.instance).expect(
+			"applying the snapshot can only fail if the passed instance is different
+			from the one that was used for creation of the snapshot;
+			we use the snapshot that is directly associated with the instance;
+			thus the snapshot was created using the instance;
+			qed",
+		);
+		f(&self.instance)
+	}
+
+	/// Returns the version of this cached runtime.
+	///
+	/// Returns `None` if the runtime doesn't provide the information or there was an error
+	/// while fetching it.
+	pub fn version(&self) -> Option<RuntimeVersion> {
+		self.version.clone()
+	}
+}
+
+/// A state snapshot of an instance taken just after instantiation.
+///
+/// It is used for restoring the state of the module after execution.
+#[derive(Clone)]
+struct StateSnapshot {
+	/// The offset and the content of the memory segments that should be used to restore the snapshot
+	data_segments: Vec<(u32, Vec<u8>)>,
+	/// The list of all global mutable variables of the module in their sequential order.
+	global_mut_values: Vec<RuntimeValue>,
+	heap_pages: u32,
+}
+
+impl StateSnapshot {
+	// Returns `None` if instance is not valid.
+	fn take(
+		module_instance: &WasmModuleInstanceRef,
+		data_segments: Vec<DataSegment>,
+		heap_pages: u32,
+	) -> Option<Self> {
+		let mut prepared_segments = Vec::with_capacity(data_segments.len());
```

Comment:
nit: This can be created without mutability:

```rust
let prepared_segments = data_segments.into_iter()
    .map(...)
    .collect::<Option<_>>()?;
```